src/librustc/mir/interpret/mod.rs

```diff
@@ -575,6 +575,144 @@ impl<Tag, Extra: Default> Allocation<Tag, Extra> {
             extra: Extra::default(),
         }
     }
+
+    #[inline]
+    pub fn size(&self) -> Size {
+        Size::from_bytes(self.bytes.len() as u64)
+    }
+
+    pub fn check_align(
+        &self,
+        offset: Size,
+        required_align: Align,
+    ) -> EvalResult<'tcx> {
+        if self.align.abi() > required_align.abi() {
+            return err!(AlignmentCheckFailed {
+                has: self.align,
+                required: required_align,
+            });
+        }
+        let offset = offset.bytes();
+        if offset % required_align.abi() == 0 {
+            Ok(())
+        } else {
+            let has = offset % required_align.abi();
+            err!(AlignmentCheckFailed {
+                has: Align::from_bytes(has, has).unwrap(),
+                required: required_align,
+            })
+        }
+    }
+
+    pub fn check_bounds(
+        &self,
+        offset: Size,
+        size: Size,
+        access: bool,
+    ) -> EvalResult<'tcx> {
+        let end = offset + size;
+        let allocation_size = self.size();
+        if end > allocation_size {
+            err!(PointerOutOfBounds { offset, access, allocation_size })
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn check_defined(
+        &self,
+        offset: Size,
+        size: Size,
+    ) -> EvalResult<'tcx> {
+        self.undef_mask.is_range_defined(
+            offset,
+            offset + size,
+        ).or_else(|idx| err!(ReadUndefBytes(idx)))
+    }
+
+    pub fn check_relocations(
```

Comment:
What exactly is this "checking"?

I spent quite some time giving all of these methods extensive docs in `memory.rs`, please don't drop these :)